api/src/license_operator.rs

```diff
@@ -0,0 +1,207 @@
+        }
+    }
+
+    #[inline(always)]
+    pub fn randomize_magic(&mut self, magic_size: usize, magic_count: usize) -> &Self {
+        for _x in 0..magic_size {
+            let mut magic: Vec<u8> = Vec::new();
+
+            info!("Randomized magic:");
+            for y in 0..magic_count {
+                let random_value = self.rng.gen();
+                magic.push(random_value);
+                info!("\t {} - {}", y, random_value);
+            }
+
+            self.magic.push(magic);
+        }
+
+        self
+    }
+
+    #[inline(always)]
+    pub fn get_serialized_key(&self, license_key: &LicenseKey) -> String {
+        self.serializer.serialize_key(&license_key.serialized_key)
+    }
+
+    #[inline(always)]
+    pub fn add_seed_to_blacklist(&mut self, seed: &[u8]) -> &Self {
+        self.blacklist.push(seed.to_vec());
+
+        self
+    }
+
+    #[inline(always)]
+    pub fn generate_license_key(&self, seed: &[u8]) -> LicenseKey {
+        // Validate user parameters
+        // Minimal 8 seed size, USER_PAYLOAD payload size and 4 checksum size
+        let license_key_required_size: usize = 8 + self.magic.payload_size() + 4;
+        if self.key_size <= license_key_required_size {
+            info!(
+                "Cannot generate license key with less than {} key size! [key_size={}]",
+                license_key_required_size, self.key_size
+            );
+            return LicenseKey::default();
+        }
+        let license_key_hash_size = self.key_size - 4 - self.magic.payload_size();
+
+        let mut license_key = LicenseKey::default();
+        let mut serialized_license_key = Vec::new();
+
+        // Hash seed to license key
+        let mut hasher = Shake256::default();
+        hasher.update(seed);
+        let mut reader = hasher.finalize_xof();
+        let mut buffer = vec![0u8; license_key_hash_size];
+        reader.read(&mut buffer);
+
+        for byte in buffer.to_vec().iter() {
+            serialized_license_key.push(*byte);
+            license_key.key.push(*byte);
+        }
+        license_key.properties.key_size = buffer.len();
+
+        // Generate payload
+        for m in self.magic.get_magic().iter() {
+            let payload = self.serializer.hash(buffer.borrow(), m);
+            serialized_license_key.push(payload);
+            license_key.payload.push(payload);
+        }
+        license_key.properties.payload_size = license_key.payload.len();
+
+        // Create checksum
+        let checksum = adler32_checksum(
+            serialized_license_key.clone(),
+            self.checksum_init.0,
+            self.checksum_init.1,
+        );
+        for byte in checksum.to_be_bytes().iter() {
+            serialized_license_key.push(*byte);
+            license_key.checksum.push(*byte);
+        }
+        license_key.properties.checksum_size = checksum.to_be_bytes().len();
+
+        license_key.serialized_key = serialized_license_key.clone();
```

Comment:
Do you actually need so many duplicated data? Because the key is already present in there somewhere. Wouldn't actually make more sense to calculate them everytime they are requested?

Reply:
Well I guess that I will have much better performance with cloning data rather than recalculating it over and over... I should make some benchmarking in future and just play with that...